signature/src/policy.rs

```diff
@@ -0,0 +1,893 @@
+        } else if self.r#type == PolicyReqType::TypeAccept.to_string() {
+            Ok(())
+        } else {
+            Err(anyhow!(ErrorInfo::ErrUnknownPrType.to_string()))
+        }
+    }
+}
+
+// The `signedBy` policy requirement:
+// The image must be signed by trusted keys for a specified identity
+#[derive(Serialize, Deserialize, Debug, PartialEq, Clone)]
+pub struct PolicyReqSignedBy {
+    r#type: String,
+
+    // KeyType specifies what kind of key reference KeyPath/KeyData is.
+    // Acceptable values are "GPGKeys" | "signedByGPGKeys” "X.509Certificates" | "signedByX.509CAs"
+    // FIXME: now only support "GPGKeys", fllowing the [containers/image](https://github.com/containers/image)
+    #[serde(rename = "keyType")]
+    key_type: String,
+
+    // KeyPath is a pathname to a local file containing the trusted key(s).
+    // Exactly one of KeyPath and KeyData must be specified.
+    #[serde(default, rename = "keyPath")]
+    key_path: String,
+    // KeyData contains the trusted key(s), base64-encoded.
+    // Exactly one of KeyPath and KeyData must be specified.
+    #[serde(default, rename = "keyData")]
+    key_data: String,
+
+    // SignedIdentity specifies what image identity the signature must be claiming about the image.
+    // Defaults to "match-exact" if not specified.
+    #[serde(default, rename = "signedIdentity")]
+    signed_identity: Option<PolicyReferenceMatch>,
+}
+
+impl PolicyReqSignedBy {
+    fn signature_judge_allowed(&self, image: &mut image::Image) -> Result<()> {
+        if self.r#type != PolicyReqType::TypeSignedBy.to_string() {
+            return Err(anyhow!(ErrorInfo::ErrUnknownPrType.to_string()));
+        }
+
+        let sigs = image.signatures()?;
+        let mut reject_reason: Vec<anyhow::Error> = Vec::new();
+
+        for sig in sigs.iter() {
+            match self.is_signature_accept(image, sig.to_vec()) {
+                // One accepted signature is enough.
+                Ok(()) => {
+                    return Ok(());
+                }
+                Err(e) => {
+                    reject_reason.push(e);
+                }
+            }
+        }
+
+        if reject_reason.is_empty() {
+            reject_reason.push(anyhow!("Can not find any signatures."));
+        }
+
+        Err(anyhow!(format!(
+            "The signatures do not satisfied! Reject reason: {:?}",
+            reject_reason
+        )))
+    }
+
+    #[allow(unused_assignments)]
+    fn is_signature_accept(&self, image: &image::Image, sig: Vec<u8>) -> Result<()> {
+        // FIXME: only support "GPGKeys" type now.
+        if self.key_type != KeyType::Gpg.to_string() {
+            return Err(anyhow!(
+                "UNKNOWN KEY TYPE in policy config: only support 'GPGKeys' now."
+            ));
+        }
+
+        if !self.key_path.is_empty() && !self.key_data.is_empty() {
+            return Err(anyhow!(
+                "BUG: both keyPath and keyData specified."
```

Comment:
The logic here is not to allow both to be specified or both to be empty, which means that the owner must explicitly specify one of them. At present, the implementation of simple signaling mechanism in golang version is so required. It may be a good idea to follow them for the time being.

Reply:
Yeah it's fine to follow the existing implementation.